### PR TITLE
rework the apache reverse proxy documentation

### PR DESCRIPTION
**Pull request by szaimen**
Signed-off-by: szaimen <szaimen@e.mail.de>

### CODE REVIEW

File: reverse-proxy.md [hunk 1]
```diff
@@ -41,14 +41,13 @@ Add this as a new Apache site config:
...
+    # Reverse proxy based on https://httpd.apache.org/docs/current/mod/mod_proxy_wstunnel.html
+    RewriteEngine on
+    ProxyPass / http://localhost:11000/
     RewriteCond %{HTTP:Upgrade} websocket [NC]
     RewriteCond %{HTTP:Connection} upgrade [NC]
-    RewriteRule ^/(.*) "ws://localhost:11000/$1" [P,L]
-    ProxyPass / http://localhost:11000/
-    ProxyPassReverse / http://localhost:11000/
+    RewriteRule ^/?(.*) "ws://localhost:11000/$1" [P,L]
+    ProxyPreserveHost On
```
Comment (szaimen) [inline reverse-proxy.md:50]: This seems to work for ws connections in general. Only Collabora still fails and I really do not understand why.

Comment (szaimen) [reply reverse-proxy.md:50]: It still fails with 

`WebSocket connection to 'wss://cloud.lindnercloud.com/cool/https%3A%2F%2Fcloud.lindnercloud.com%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F4_oc7k52n1zdg3%3Faccess_token%3DGFt6n7UK5HGxa83XTVo0g4fegz6kvzAB%26access_token_ttl%3D0%26permission%3Dedit/ws?WOPISrc=https%3A%2F%2Fcloud.lindnercloud.com%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F4_oc7k52n1zdg3&compat=/ws' failed:`

File: Containers/apache/Caddyfile [hunk 2]
```diff
@@ -28,16 +28,19 @@
     route /browser/* {
         reverse_proxy {$COLLABORA_HOST}:9980 {
             # trusted_proxies placeholder
+            header_up Host {http.request.host}
```
Comment (szaimen) [inline Containers/apache/Caddyfile:31]: Not sure if this is really needed aa it doesnt make it work anyway. Question would then also be why it currently works even without it

**Comment by Zoey2936** [comment]
Bellow logs of Collabora Container:

`AllowEncodedSlashes On`:
```
wsd-00001-00039 2022-10-08 21:47:24.386941 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/lc_insertcolumnsafter.svg].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:24.387563 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/lc_insertrowsafter.svg].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:24.389069 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/lc_entirerow.svg].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:24.496987 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/welcome/slide2.png].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:24.497122 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/welcome/slide3.png].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:24.512319 +0200 [ websrv_poll ] ERR  FileServerRequestHandler: File not found: Invalid URI request: [/browser/abd9205/images/welcome/slide1-left.png].| wsd/FileServer.cpp:692

wsd-00001-00039 2022-10-08 21:47:51.547051 +0200 [ websrv_poll ] ERR  #32 Exception while processing incoming request: [GET /cool/https:/cloud.scdmx.de/index.php/apps/richdocuments/wopi/files/3924_oc1iyepb67f2?access_token=6XSM1G4jpk77Y7jxnEBRChgQP3keTA6m&access_token_ttl=0/ws HTTP/1.1
...]: Bad URI syntax| wsd/COOLWSD.cpp:3757

wsd-00001-00039 2022-10-08 21:47:52.135610 +0200 [ websrv_poll ] ERR  #32 Exception while processing incoming request: [GET /cool/https:/cloud.scdmx.de/index.php/apps/richdocuments/wopi/files/3924_oc1iyepb67f2?access_token=6XSM1G4jpk77Y7jxnEBRChgQP3keTA6m&access_token_ttl=0&permission=edit/ws HTTP/1.1
...]: Bad URI syntax| wsd/COOLWSD.cpp:3757

wsd-00001-00039 2022-10-08 21:47:54.249145 +0200 [ websrv_poll ] ERR  #32 Exception while processing incoming request: [GET /cool/https:/cloud.scdmx.de/index.php/apps/richdocuments/wopi/files/3924_oc1iyepb67f2?access_token=6XSM1G4jpk77Y7jxnEBRChgQP3keTA6m&access_token_ttl=0&permission=edit/ws HTTP/1.1
...]: Bad URI syntax| wsd/COOLWSD.cpp:3757
```

`AllowEncodedSlashes NoDecode`:
```
wsd-00001-00039 2022-10-08 21:45:50.625100 +0200 [ websrv_poll ] ERR  Unknown resource: /cool/https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2?access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0/ws GWebSocket, host: cloud.scdmx.de, path: 4

[0] 'cool'

[1] 'https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2'

[2] 'access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0'

[3] 'ws'

full URI: /cool/https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2?access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0/ws| wsd/COOLWSD.cpp:3747

wsd-00001-00039 2022-10-08 21:45:53.165426 +0200 [ websrv_poll ] ERR  Unknown resource: /cool/https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2?access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0&permission=edit/ws GWebSocket, host: cloud.scdmx.de, path: 4

[0] 'cool'

[1] 'https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2'

[2] 'access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0&permission=edit'

[3] 'ws'

full URI: /cool/https:%252F%252Fcloud.scdmx.de%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F3924_oc1iyepb67f2?access_token=38IR6jHZdwZWqcucLPsY3Thez66JGeri&access_token_ttl=0&permission=edit/ws| wsd/COOLWSD.cpp:3747
```

No option:
 - empty collabora logs

File: reverse-proxy.md [hunk 3]
```diff
@@ -41,14 +41,13 @@ Add this as a new Apache site config:
 <VirtualHost *:443>
     ServerName <your-nc-domain>
 
-    # Reverse proxy
-    RewriteEngine On
-    ProxyPreserveHost On
+    # Reverse proxy based on https://httpd.apache.org/docs/current/mod/mod_proxy_wstunnel.html
+    RewriteEngine on
+    ProxyPass / http://localhost:11000/
```
Comment (szaimen) [inline reverse-proxy.md:46]: Could try the nocanon option here. Also try the encoded slashes option

File: reverse-proxy.md (hunk 1 above)
Comment (szaimen) [reply reverse-proxy.md:50]: Maybe try off here of this option

Comment (szaimen) [reply reverse-proxy.md:50]: Also ProxyRequests off

**Comment by szaimen** [comment]
@Zoey2936 due to your help, I am now up to something, I think. as the allowencodedslashes nodecode allows to log the complete url in the collabora logs.

This is how the url looks on a working instance:
`/cool/https%3A%2F%2Faio-testing.mydomain.com%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F25_ocdph6m977ya%3Faccess_token%3D6PmFoHyEV3U8bHJMSrBPqBtSbkaPOggO%26access_token_ttl%3D0/ws?WOPISrc=https%3A%2F%2Faio-testing.mydomain.com%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F25_ocdph6m977ya&compat=/ws

and this is how it looks on a not working instance behind apache reverse proxy:
`/cool/https:%252F%252Fcloud.mydomain.com%252Findex.php%252Fapps%252Frichdocuments%252Fwopi%252Ffiles%252F4_oc7k52n1zdg3?access_token=wKnBgYjeV1DWPrpa8EParrZGzkWDCo7v&access_token_ttl=0/ws&WOPISrc=https%3A%2F%2Fcloud.mydomain.com%2Findex.php%2Fapps%2Frichdocuments%2Fwopi%2Ffiles%2F4_oc7k52n1zdg3&compat=/ws`

Do you have any idea what could break the encoding and/or how to fix it?

File: reverse-proxy.md [hunk 4]
```diff
@@ -41,14 +41,14 @@ Add this as a new Apache site config:
...
+    RewriteEngine on
     ProxyPreserveHost On
+    ProxyPass / http://localhost:11000/
     RewriteCond %{HTTP:Upgrade} websocket [NC]
     RewriteCond %{HTTP:Connection} upgrade [NC]
-    RewriteRule ^/(.*) "ws://localhost:11000/$1" [P,L]
-    ProxyPass / http://localhost:11000/
-    ProxyPassReverse / http://localhost:11000/
+    RewriteRule ^/?(.*) "ws://localhost:11000/$1" [P,L,QSA]
+    AllowEncodedSlashes NoDecode
```
Comment (szaimen) [inline reverse-proxy.md:51]: should most likely get removed before merging as it is only for debugging.

Comment (szaimen) [reply reverse-proxy.md]: no, is needed

**Comment by szaimen** [comment]
Thanks a lot @Zoey2936! AllowEncodedSlashes NoDecode was part of the solution!

**Comment by szaimen** [comment]
This works now 🎉🎉🎉

**Comment by TheManchineel** [comment]
> Thanks a lot @Zoey2936! AllowEncodedSlashes NoDecode was part of the solution!

THANK YOU THANK YOU THANK YOU! The equivalent of adding a `AllowEncodedSlashes NoDecode` clause on NGINX is to replace the `proxy_pass http://server_hostname:11000/` directive with `proxy_pass http://server_hostname:11000$request_uri`. This solved the same issue for me on NGINX!

**Pull request merged**